File: platforms/ios/Sources/Feedback/Survey/SurveyPanel.swift
#if canImport(UIKit)
import UIKit

/**
 Microsurvey sheet.

 A bottom sheet rather than a full-screen takeover: a research survey interrupts
 someone who opened the app to do something else, and a takeover converts
 curiosity into dismissal. One question at a time keeps it small enough to
 answer without feeling like a form.

 Partial responses are returned, not discarded. Someone who answers two of four
 questions and closes has still told the researcher something, and dropping it
 would bias results toward people with time to finish.

 Mirrors `SurveyPanel.kt`.
 */
final class SurveyPanel: UIView {

    private let study: Study
    private let onDone: ([Answer], Bool, Int64) -> Void
    private let onDismiss: () -> Void

    private var index = 0
    private var drafts: [String: AnswerDraft] = [:]
    private let startedAt = Date()

    private let sheet = UIView()
    private let bodyStack = UIStackView()
    private let progressFill = UIView()
    private var progressWidth: NSLayoutConstraint?
    private let counter = UILabel()
    private let nextButton = UIButton(type: .system)
    private let skipButton = UIButton(type: .system)

    private let accent = UIColor(red: 0.31, green: 0.27, blue: 0.90, alpha: 1)
    private let border = UIColor(red: 0.89, green: 0.91, blue: 0.93, alpha: 1)
    private let muted = UIColor(red: 0.40, green: 0.44, blue: 0.52, alpha: 1)

    init(
        study: Study,
        onDone: @escaping ([Answer], Bool, Int64) -> Void,
        onDismiss: @escaping () -> Void
    ) {
        self.study = study
        self.onDone = onDone
        self.onDismiss = onDismiss
        super.init(frame: .zero)

        backgroundColor = UIColor.black.withAlphaComponent(0.4)
        buildLayout()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported; this view is created in code")
    }

    /// Presents the sheet, or reports a dismissal if the study is malformed.
    func present(in window: UIWindow) -> Bool {
        let problems = validateStudy(study)
        guard problems.isEmpty else {
            // Loudly at integration time, rather than a blank question shown to
            // a real user.
            print("[markerio] invalid study \"\(study.id)\":\n  \(problems.joined(separator: "\n  "))")
            onDismiss()
            return false
        }

        frame = window.bounds
        autoresizingMask = [.flexibleWidth, .flexibleHeight]
        window.addSubview(self)
        renderStep()
        return true
    }

    private func finish(completed: Bool) {
        let answers = study.questions.compactMap { question -> Answer? in
            guard let draft = drafts[question.id] else { return nil }
            return toAnswer(questionId: question.id, draft: draft)
        }

        removeFromSuperview()

        // Nothing answered and dismissed: a decline, not a response.
        if !completed && answers.isEmpty {
            onDismiss()
            return
        }
        onDone(answers, completed, Int64(Date().timeIntervalSince(startedAt) * 1000))
    }

    // MARK: - chrome

    private func buildLayout() {
        // Tapping the scrim dismisses, matching platform convention for a sheet
        // the user did not ask for.
        let scrimTap = UITapGestureRecognizer(target: self, action: #selector(scrimTapped))
        addGestureRecognizer(scrimTap)

        sheet.backgroundColor = .white
        sheet.layer.cornerRadius = 16
        sheet.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
        sheet.translatesAutoresizingMaskIntoConstraints = false
        // Swallow taps so they do not reach the dismissing scrim behind.
        sheet.addGestureRecognizer(UITapGestureRecognizer(target: nil, action: nil))
        addSubview(sheet)

        let track = UIView()
        track.backgroundColor = UIColor(red: 0.95, green: 0.96, blue: 0.97, alpha: 1)
        track.layer.cornerRadius = 2
        track.translatesAutoresizingMaskIntoConstraints = false

        progressFill.backgroundColor = accent
        progressFill.layer.cornerRadius = 2
        progressFill.translatesAutoresizingMaskIntoConstraints = false
        track.addSubview(progressFill)

        let close = UIButton(type: .system)
        close.setTitle("✕", for: .normal)
        close.setTitleColor(muted, for: .normal)
        close.accessibilityIdentifier = "survey-close"
        close.addTarget(self, action: #selector(closeTapped), for: .touchUpInside)
        close.translatesAutoresizingMaskIntoConstraints = false

        bodyStack.axis = .vertical
        bodyStack.spacing = 10
        bodyStack.translatesAutoresizingMaskIntoConstraints = false

        counter.font = .systemFont(ofSize: 12)
        counter.textColor = UIColor(red: 0.60, green: 0.64, blue: 0.70, alpha: 1)

        skipButton.setTitle("Skip", for: .normal)
        skipButton.setTitleColor(muted, for: .normal)
        skipButton.titleLabel?.font = .systemFont(ofSize: 13, weight: .medium)
        skipButton.accessibilityIdentifier = "survey-skip"
        skipButton.addTarget(self, action: #selector(skipTapped), for: .touchUpInside)

        nextButton.setTitleColor(.white, for: .normal)
        nextButton.titleLabel?.font = .systemFont(ofSize: 13, weight: .semibold)
        nextButton.backgroundColor = accent
        nextButton.layer.cornerRadius = 7
        nextButton.contentEdgeInsets = UIEdgeInsets(top: 9, left: 18, bottom: 9, right: 18)
        nextButton.accessibilityIdentifier = "survey-next"
        nextButton.addTarget(self, action: #selector(nextTapped), for: .touchUpInside)

        let footer = UIStackView(arrangedSubviews: [counter, skipButton, nextButton])
        footer.axis = .horizontal
        footer.spacing = 8
        footer.alignment = .center
        footer.translatesAutoresizingMaskIntoConstraints = false

        sheet.addSubview(track)
        sheet.addSubview(close)
        sheet.addSubview(bodyStack)
        sheet.addSubview(footer)

        let fillWidth = progressFill.widthAnchor.constraint(equalToConstant: 0)
        progressWidth = fillWidth

        NSLayoutConstraint.activate([
            sheet.leadingAnchor.constraint(equalTo: leadingAnchor),
            sheet.trailingAnchor.constraint(equalTo: trailingAnchor),
            // safeAreaLayoutGuide, or the footer sits under the home indicator.
            sheet.bottomAnchor.constraint(equalTo: bottomAnchor),

            track.topAnchor.constraint(equalTo: sheet.topAnchor, constant: 14),
            track.leadingAnchor.constraint(equalTo: sheet.leadingAnchor, constant: 20),
            track.heightAnchor.constraint(equalToConstant: 3),
            track.trailingAnchor.constraint(equalTo: close.leadingAnchor, constant: -10),

            progressFill.leadingAnchor.constraint(equalTo: track.leadingAnchor),
            progressFill.topAnchor.constraint(equalTo: track.topAnchor),
            progressFill.bottomAnchor.constraint(equalTo: track.bottomAnchor),
            fillWidth,

            close.trailingAnchor.constraint(equalTo: sheet.trailingAnchor, constant: -16),
            close.centerYAnchor.constraint(equalTo: track.centerYAnchor),

            bodyStack.topAnchor.constraint(equalTo: track.bottomAnchor, constant: 18),
            bodyStack.leadingAnchor.constraint(equalTo: sheet.leadingAnchor, constant: 20),
            bodyStack.trailingAnchor.constraint(equalTo: sheet.trailingAnchor, constant: -20),

            footer.topAnchor.constraint(equalTo: bodyStack.bottomAnchor, constant: 18),
            footer.leadingAnchor.constraint(equalTo: sheet.leadingAnchor, constant: 20),
            footer.trailingAnchor.constraint(equalTo: sheet.trailingAnchor, constant: -20),
            footer.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor, constant: -18),
        ])
    }

    @objc private func scrimTapped() { finish(completed: false) }
    @objc private func closeTapped() { finish(completed: false) }
    @objc private func skipTapped() { advance(skipped: true) }
    @objc private func nextTapped() { advance(skipped: false) }

    // MARK: - steps

    private func renderStep() {
        guard index < study.questions.count else { return }
        let question = study.questions[index]
        let total = study.questions.count

        counter.text = "\(index + 1) of \(total)"
        nextButton.setTitle(index == total - 1 ? "Submit" : "Next", for: .normal)
        // A required question cannot be skipped, so the affordance is removed
        // rather than left visible and inert.
        skipButton.isHidden = question.isRequired

        layoutIfNeeded()
        progressWidth?.constant = (bounds.width - 40) * CGFloat(index) / CGFloat(total)

        bodyStack.arrangedSubviews.forEach { $0.removeFromSuperview() }

        let prompt = UILabel()
        prompt.text = question.prompt
        prompt.font = .systemFont(ofSize: 16, weight: .semibold)
        prompt.numberOfLines = 0
        bodyStack.addArrangedSubview(prompt)

        if let help = question.help {
            let label = UILabel()
            label.text = help
            label.font = .systemFont(ofSize: 13)
            label.textColor = muted
            label.numberOfLines = 0
            bodyStack.addArrangedSubview(label)
        }

        switch question {
        case let .nps(id, _, _, _, labels):
            bodyStack.addArrangedSubview(scaleView(id: id, from: 0, to: 10, labels: labels))
        case let .rating(id, _, _, _, scale, labels):
            bodyStack.addArrangedSubview(scaleView(id: id, from: 1, to: scale, labels: labels))
        case let .choice(id, _, _, _, options, multiple):
            bodyStack.addArrangedSubview(choiceView(id: id, options: options, multiple: multiple))
        case let .text(id, _, _, _, placeholder, maxLength):
            bodyStack.addArrangedSubview(textView(id: id, placeholder: placeholder, maxLength: maxLength))
        }

        syncNext()
    }

    private var scaleButtons: [Int: UIButton] = [:]

    private func scaleView(id: String, from: Int, to: Int, labels: (String, String)?) -> UIView {
        scaleButtons.removeAll()

        let row = UIStackView()
        row.axis = .horizontal
        row.distribution = .fillEqually
        row.spacing = 3

        for value in from...to {
            let button = UIButton(type: .system)
            button.setTitle(String(value), for: .normal)
            button.titleLabel?.font = .systemFont(ofSize: 13, weight: .medium)
            button.layer.cornerRadius = 7
            button.layer.borderWidth = 1
            button.accessibilityIdentifier = "survey-scale-\(value)"
            button.tag = value
            button.addTarget(self, action: #selector(scaleTapped(_:)), for: .touchUpInside)
            button.heightAnchor.constraint(equalToConstant: 40).isActive = true

            scaleButtons[value] = button
            row.addArrangedSubview(button)
        }

        scaleQuestionId = id
        styleScale()

        guard let labels else { return row }

        let low = UILabel()
        low.text = labels.0
        low.font = .systemFont(ofSize: 11)
        low.textColor = UIColor(red: 0.60, green: 0.64, blue: 0.70, alpha: 1)

        let high = UILabel()
        high.text = labels.1
        high.font = .systemFont(ofSize: 11)
        high.textColor = low.textColor
        high.textAlignment = .right

        let ends = UIStackView(arrangedSubviews: [low, high])
        ends.axis = .horizontal
        ends.distribution = .fillEqually

        let column = UIStackView(arrangedSubviews: [row, ends])
        column.axis = .vertical
        column.spacing = 6
        return column
    }

    private var scaleQuestionId = ""

    @objc private func scaleTapped(_ sender: UIButton) {
        drafts[scaleQuestionId] = .number(sender.tag)
        styleScale()
        syncNext()
    }

    private func styleScale() {
        let selected: Int? = {
            if case let .number(value)? = drafts[scaleQuestionId] { return value }
            return nil
        }()

        for (value, button) in scaleButtons {
            let on = value == selected
            button.backgroundColor = on ? accent : .clear
            button.layer.borderColor = (on ? accent : border).cgColor
            button.setTitleColor(on ? .white : UIColor(red: 0.28, green: 0.33, blue: 0.40, alpha: 1), for: .normal)
        }
    }

    private var choiceButtons: [String: UIButton] = [:]
    private var choiceQuestionId = ""
    private var choiceMultiple = false

    private func choiceView(id: String, options: [String], multiple: Bool) -> UIView {
        choiceButtons.removeAll()
        choiceQuestionId = id
        choiceMultiple = multiple

        let column = UIStackView()
        column.axis = .vertical
        column.spacing = 6

        for option in options {
            let button = UIButton(type: .system)
            button.setTitle(option, for: .normal)
            button.titleLabel?.font = .systemFont(ofSize: 14)
            button.contentHorizontalAlignment = .left
            button.contentEdgeInsets = UIEdgeInsets(top: 11, left: 12, bottom: 11, right: 12)
            button.layer.cornerRadius = 8
            button.layer.borderWidth = 1
            button.accessibilityIdentifier = "survey-choice-\(option)"
            button.addTarget(self, action: #selector(choiceTapped(_:)), for: .touchUpInside)

            choiceButtons[option] = button
            column.addArrangedSubview(button)
        }

        styleChoices()
        return column
    }

    @objc private func choiceTapped(_ sender: UIButton) {
        guard let option = sender.title(for: .normal) else { return }

        if choiceMultiple {
            var current: [String] = {
                if case let .multiple(values)? = drafts[choiceQuestionId] { return values }
                return []
            }()

            if let at = current.firstIndex(of: option) { current.remove(at: at) } else { current.append(option) }
            if current.isEmpty { drafts.removeValue(forKey: choiceQuestionId) }
            else { drafts[choiceQuestionId] = .multiple(current) }
        } else {
            drafts[choiceQuestionId] = .single(option)
        }

        styleChoices()
        syncNext()
    }

    private func styleChoices() {
        let selected: Set<String> = {
            switch drafts[choiceQuestionId] {
            case let .multiple(values): return Set(values)
            case let .single(value): return [value]
            default: return []
            }
        }()

        for (option, button) in choiceButtons {
            let on = selected.contains(option)
            button.backgroundColor = on ? UIColor(red: 0.93, green: 0.95, blue: 1, alpha: 1) : .clear
            button.layer.borderColor = (on ? accent : border).cgColor
            button.setTitleColor(on ? UIColor(red: 0.21, green: 0.19, blue: 0.64, alpha: 1)
                                    : UIColor(red: 0.20, green: 0.25, blue: 0.33, alpha: 1), for: .normal)
        }
    }

    private var textQuestionId = ""

    private func textView(id: String, placeholder: String?, maxLength: Int) -> UIView {
        textQuestionId = id

        let field = UITextView()
        field.font = .systemFont(ofSize: 14)
        field.backgroundColor = UIColor(red: 0.98, green: 0.98, blue: 0.98, alpha: 1)
        field.layer.cornerRadius = 8
        field.layer.borderWidth = 1
        field.layer.borderColor = border.cgColor
        field.textContainerInset = UIEdgeInsets(top: 10, left: 8, bottom: 10, right: 8)
        field.accessibilityIdentifier = "survey-text"
        field.delegate = self
        field.heightAnchor.constraint(equalToConstant: 84).isActive = true

        if case let .free(existing)? = drafts[id] { field.text = existing }
        return field
    }

    private func syncNext() {
        guard index < study.questions.count else { return }
        let question = study.questions[index]
        let ok = !question.isRequired || isAnswered(question, drafts[question.id])

        nextButton.isEnabled = ok
        nextButton.alpha = ok ? 1 : 0.45
    }

    private func advance(skipped: Bool) {
        if skipped, index < study.questions.count {
            drafts.removeValue(forKey: study.questions[index].id)
        }

        guard index < study.questions.count - 1 else {
            showThanks()
            return
        }
        index += 1
        renderStep()
    }

    private func showThanks() {
        bodyStack.arrangedSubviews.forEach { $0.removeFromSuperview() }

        let label = UILabel()
        label.text = study.thanks ?? "Thanks — that helps."
        label.font = .systemFont(ofSize: 15)
        label.textColor = muted
        label.textAlignment = .center
        label.numberOfLines = 0
        bodyStack.addArrangedSubview(label)

        counter.isHidden = true
        skipButton.isHidden = true
        nextButton.isHidden = true

        // Resolve now so the host records the response immediately; the sheet
        // lingers only as an acknowledgement.
        let answers = study.questions.compactMap { question -> Answer? in
            guard let draft = drafts[question.id] else { return nil }
            return toAnswer(questionId: question.id, draft: draft)
        }
        onDone(answers, true, Int64(Date().timeIntervalSince(startedAt) * 1000))

        DispatchQueue.main.asyncAfter(deadline: .now() + 1.6) { [weak self] in
            self?.removeFromSuperview()
        }
    }
}

extension SurveyPanel: UITextViewDelegate {
    func textViewDidChange(_ textView: UITextView) {
        let value = textView.text ?? ""
        if value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            drafts.removeValue(forKey: textQuestionId)
        } else {
            drafts[textQuestionId] = .free(value)
        }
        syncNext()
    }
}
#endif

File: platforms/android/feedback/src/main/kotlin/io/markerusa/feedback/survey/SurveyPanel.kt
package io.markerusa.feedback.survey

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.text.InputType
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import io.markerusa.feedback.Answer

/**
 * Microsurvey panel.
 *
 * A bottom sheet rather than a full-screen takeover: a research survey
 * interrupts someone who opened the app to do something else, and a takeover
 * converts curiosity into dismissal. One question at a time keeps it small
 * enough to answer without feeling like a form.
 *
 * Partial responses are returned, not discarded. Someone who answers two of four
 * questions and closes has still told the researcher something, and dropping it
 * would bias results toward people with time to finish.
 *
 * Built in code rather than XML: a library shipping layout resources forces
 * resource merging on every host app and risks id collisions.
 */
internal class SurveyPanel(
    private val activity: Activity,
    private val study: Study,
    private val onDone: (answers: List<Answer>, completed: Boolean, durationMs: Long) -> Unit,
    private val onDismiss: () -> Unit
) {
    private companion object {
        const val ACCENT = "#4F46E5"
        const val SURFACE = 0xFFFFFFFF.toInt()
        const val TEXT = 0xFF101828.toInt()
        const val MUTED = 0xFF667085.toInt()
        const val BORDER = 0xFFE4E7EC.toInt()
        const val FIELD_BG = 0xFFF9FAFB.toInt()
    }

    private var root: FrameLayout? = null
    private var index = 0
    private val values = LinkedHashMap<String, Any>()
    private val startedAt = System.currentTimeMillis()

    private lateinit var body: LinearLayout
    private lateinit var progressFill: View
    private lateinit var counter: TextView
    private lateinit var nextButton: Button
    private lateinit var skipButton: Button

    private val density = activity.resources.displayMetrics.density
    private fun dp(value: Int) = (value * density).toInt()

    fun show() {
        val problems = validateStudy(study)
        if (problems.isNotEmpty()) {
            // Loudly at integration time, rather than a blank question shown to
            // a real user.
            android.util.Log.e("MarkerFeedback", "invalid study \"${study.id}\":\n  ${problems.joinToString("\n  ")}")
            onDismiss()
            return
        }

        val content = activity.findViewById<ViewGroup>(android.R.id.content)

        val scrim = FrameLayout(activity).apply {
            setBackgroundColor(0x66000000)
            isClickable = true
            isFocusable = true
            // Tapping the scrim dismisses, matching platform convention for a
            // sheet the user did not ask for.
            setOnClickListener { finish(false) }
        }

        val sheet = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            background = GradientDrawable().apply {
                cornerRadii = floatArrayOf(
                    dp(16).toFloat(), dp(16).toFloat(), dp(16).toFloat(), dp(16).toFloat(),
                    0f, 0f, 0f, 0f
                )
                setColor(SURFACE)
            }
            setPadding(dp(20), dp(14), dp(20), dp(18))
            // Swallow taps so they do not reach the dismissing scrim behind.
            isClickable = true
        }

        sheet.addView(buildHeader())
        body = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
        sheet.addView(body, LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(14) })
        sheet.addView(buildFooter(), LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(16) })

        scrim.addView(sheet, FrameLayout.LayoutParams(MATCH, WRAP, Gravity.BOTTOM))
        content.addView(scrim, ViewGroup.LayoutParams(MATCH, MATCH))
        root = scrim

        renderStep()
    }

    private fun dismissView() {
        root?.let { (it.parent as? ViewGroup)?.removeView(it) }
        root = null
    }

    private fun finish(completed: Boolean) {
        val answers = values.map { (questionId, value) -> toAnswer(questionId, value) }
        dismissView()

        // Nothing answered and dismissed: a decline, not a response.
        if (!completed && answers.isEmpty()) {
            onDismiss()
            return
        }
        onDone(answers, completed, System.currentTimeMillis() - startedAt)
    }

    // --- chrome ---------------------------------------------------------------

    private fun buildHeader(): View {
        val row = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
        }

        val track = FrameLayout(activity).apply {
            background = GradientDrawable().apply {
                cornerRadius = dp(2).toFloat()
                setColor(0xFFF2F4F7.toInt())
            }
        }
        progressFill = View(activity).apply {
            background = GradientDrawable().apply {
                cornerRadius = dp(2).toFloat()
                setColor(Color.parseColor(ACCENT))
            }
        }
        track.addView(progressFill, FrameLayout.LayoutParams(0, dp(3)))

        val close = TextView(activity).apply {
            text = "✕"
            textSize = 15f
            setTextColor(MUTED)
            setPadding(dp(12), dp(4), dp(4), dp(4))
            contentDescription = "Dismiss survey"
            setOnClickListener { finish(false) }
        }

        row.addView(track, LinearLayout.LayoutParams(0, dp(3)).apply { weight = 1f })
        row.addView(close)
        return row
    }

    private fun buildFooter(): View {
        val row = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
        }

        counter = TextView(activity).apply {
            textSize = 12f
            setTextColor(0xFF98A2B3.toInt())
        }

        skipButton = Button(activity).apply {
            text = "Skip"
            isAllCaps = false
            textSize = 13f
            setTextColor(MUTED)
            background = GradientDrawable().apply {
                cornerRadius = dp(7).toFloat()
                setColor(Color.TRANSPARENT)
            }
            setOnClickListener { advance(skipped = true) }
        }

        nextButton = Button(activity).apply {
            isAllCaps = false
            textSize = 13f
            setTextColor(Color.WHITE)
            background = GradientDrawable().apply {
                cornerRadius = dp(7).toFloat()
                setColor(Color.parseColor(ACCENT))
            }
            setPadding(dp(18), 0, dp(18), 0)
            setOnClickListener { advance(skipped = false) }
        }

        row.addView(counter, LinearLayout.LayoutParams(0, WRAP).apply { weight = 1f })
        row.addView(skipButton)
        row.addView(nextButton, LinearLayout.LayoutParams(WRAP, WRAP).apply { marginStart = dp(6) })
        return row
    }

    // --- steps ----------------------------------------------------------------

    private fun renderStep() {
        val question = study.questions.getOrNull(index) ?: return
        val total = study.questions.size

        counter.text = "${index + 1} of $total"
        nextButton.text = if (index == total - 1) "Submit" else "Next"
        // A required question cannot be skipped, so the affordance is removed
        // rather than left visible and inert.
        skipButton.visibility = if (question.required) View.GONE else View.VISIBLE

        progressFill.layoutParams = (progressFill.layoutParams as FrameLayout.LayoutParams).apply {
            width = ((activity.resources.displayMetrics.widthPixels - dp(40)) * index / total)
        }
        progressFill.requestLayout()

        body.removeAllViews()
        body.addView(TextView(activity).apply {
            text = question.prompt
            textSize = 16f
            setTextColor(TEXT)
            setTypeface(typeface, android.graphics.Typeface.BOLD)
        })

        question.help?.let {
            body.addView(TextView(activity).apply {
                text = it
                textSize = 13f
                setTextColor(MUTED)
            }, LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(4) })
        }

        val input = when (question) {
            is Question.Nps -> scaleView(question.id, 0, 10, question.labels)
            is Question.Rating -> scaleView(question.id, 1, question.scale, question.labels)
            is Question.Choice -> choiceView(question)
            is Question.Text -> textView(question)
        }
        body.addView(input, LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(14) })

        syncNext()
    }

    private fun scaleView(
        questionId: String,
        from: Int,
        to: Int,
        labels: Pair<String, String>?
    ): View {
        val column = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
        val row = LinearLayout(activity).apply { orientation = LinearLayout.HORIZONTAL }
        val buttons = mutableMapOf<Int, Button>()

        fun style(button: Button, on: Boolean) {
            button.background = GradientDrawable().apply {
                cornerRadius = dp(7).toFloat()
                setColor(if (on) Color.parseColor(ACCENT) else Color.TRANSPARENT)
                setStroke(dp(1), if (on) Color.parseColor(ACCENT) else BORDER)
            }
            button.setTextColor(if (on) Color.WHITE else 0xFF475467.toInt())
        }

        for (value in from..to) {
            val button = Button(activity).apply {
                text = value.toString()
                textSize = 13f
                isAllCaps = false
                setPadding(0, 0, 0, 0)
                minWidth = 0
                minimumWidth = 0
                setOnClickListener {
                    values[questionId] = value
                    buttons.forEach { (v, b) -> style(b, v == value) }
                    syncNext()
                }
            }
            style(button, values[questionId] == value)
            buttons[value] = button
            row.addView(button, LinearLayout.LayoutParams(0, dp(40)).apply {
                weight = 1f
                marginEnd = dp(3)
            })
        }

        column.addView(row)

        labels?.let { (low, high) ->
            val ends = LinearLayout(activity).apply { orientation = LinearLayout.HORIZONTAL }
            ends.addView(TextView(activity).apply {
                text = low; textSize = 11f; setTextColor(0xFF98A2B3.toInt())
            }, LinearLayout.LayoutParams(0, WRAP).apply { weight = 1f })
            ends.addView(TextView(activity).apply {
                text = high; textSize = 11f; setTextColor(0xFF98A2B3.toInt()); gravity = Gravity.END
            }, LinearLayout.LayoutParams(0, WRAP).apply { weight = 1f })
            column.addView(ends, LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(6) })
        }

        return column
    }

    private fun choiceView(question: Question.Choice): View {
        val column = LinearLayout(activity).apply { orientation = LinearLayout.VERTICAL }
        val buttons = mutableMapOf<String, Button>()

        fun style(button: Button, on: Boolean) {
            button.background = GradientDrawable().apply {
                cornerRadius = dp(8).toFloat()
                setColor(if (on) 0xFFEEF2FF.toInt() else Color.TRANSPARENT)
                setStroke(dp(1), if (on) Color.parseColor(ACCENT) else BORDER)
            }
            button.setTextColor(if (on) 0xFF3730A3.toInt() else 0xFF344054.toInt())
        }

        for (option in question.options) {
            val button = Button(activity).apply {
                text = option
                textSize = 14f
                isAllCaps = false
                gravity = Gravity.START or Gravity.CENTER_VERTICAL
                setPadding(dp(12), dp(10), dp(12), dp(10))
                setOnClickListener {
                    if (question.multiple) {
                        @Suppress("UNCHECKED_CAST")
                        val current = (values[question.id] as? List<String>)?.toMutableList() ?: mutableListOf()
                        if (current.contains(option)) current.remove(option) else current.add(option)

                        if (current.isEmpty()) values.remove(question.id) else values[question.id] = current
                        style(this, current.contains(option))
                    } else {
                        values[question.id] = option
                        buttons.forEach { (o, b) -> style(b, o == option) }
                    }
                    syncNext()
                }
            }

            val selected = if (question.multiple) {
                @Suppress("UNCHECKED_CAST")
                (values[question.id] as? List<String>)?.contains(option) == true
            } else {
                values[question.id] == option
            }
            style(button, selected)

            buttons[option] = button
            column.addView(button, LinearLayout.LayoutParams(MATCH, WRAP).apply { topMargin = dp(6) })
        }

        return column
    }

    private fun textView(question: Question.Text): View =
        EditText(activity).apply {
            hint = question.placeholder ?: "Type your answer…"
            textSize = 14f
            setTextColor(TEXT)
            setHintTextColor(0xFF98A2B3.toInt())
            background = GradientDrawable().apply {
                cornerRadius = dp(8).toFloat()
                setColor(FIELD_BG)
                setStroke(dp(1), BORDER)
            }
            setPadding(dp(11), dp(10), dp(11), dp(10))
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE
            minLines = 3
            gravity = Gravity.TOP or Gravity.START
            filters = arrayOf(android.text.InputFilter.LengthFilter(question.maxLength))
            setText(values[question.id] as? String ?: "")

            addTextChangedListener(object : android.text.TextWatcher {
                override fun afterTextChanged(s: android.text.Editable?) {
                    val value = s?.toString().orEmpty()
                    if (value.isBlank()) values.remove(question.id) else values[question.id] = value
                    syncNext()
                }
                override fun beforeTextChanged(s: CharSequence?, a: Int, b: Int, c: Int) {}
                override fun onTextChanged(s: CharSequence?, a: Int, b: Int, c: Int) {}
            })
        }

    private fun syncNext() {
        val question = study.questions.getOrNull(index) ?: return
        val ok = !question.required || isAnswered(question, values[question.id])
        nextButton.isEnabled = ok
        nextButton.alpha = if (ok) 1f else 0.45f
    }

    private fun advance(skipped: Boolean) {
        val question = study.questions.getOrNull(index)
        if (question != null && skipped) values.remove(question.id)

        if (index >= study.questions.size - 1) {
            showThanks()
            return
        }
        index++
        renderStep()
    }

    private fun showThanks() {
        body.removeAllViews()
        body.addView(TextView(activity).apply {
            text = study.thanks ?: "Thanks — that helps."
            textSize = 15f
            setTextColor(0xFF475467.toInt())
            gravity = Gravity.CENTER
            setPadding(0, dp(18), 0, dp(18))
        })
        counter.visibility = View.GONE
        skipButton.visibility = View.GONE
        nextButton.visibility = View.GONE

        // Resolve now so the host records the response immediately; the sheet
        // lingers only as an acknowledgement.
        val answers = values.map { (questionId, value) -> toAnswer(questionId, value) }
        onDone(answers, true, System.currentTimeMillis() - startedAt)

        body.postDelayed({ dismissView() }, 1600)
    }
}

private const val MATCH = ViewGroup.LayoutParams.MATCH_PARENT
private const val WRAP = ViewGroup.LayoutParams.WRAP_CONTENT
